usersession/userd/launcher.go

```diff
@@ -183,12 +189,190 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+	if !strings.HasPrefix(desktopFile, dirs.SnapDesktopFilesDir+"/") {
+		// We currently only support launching snap applications from desktop files in
+		// /var/lib/snapd/desktop/applications and these desktop files are written by snapd and
+		// considered safe for userd to process. If other directories are added in the future,
+		// verifyDesktopFileLocation() and parseExecCommand() may need to be updated.
+		return fmt.Errorf("only launching snap applications from /var/lib/snapd/desktop/applications is supported")
+	}
+
+	if filepath.Clean(desktopFile) != desktopFile {
+		return fmt.Errorf("desktop file has unclean path: %q", desktopFile)
+	}
+
+	return nil
+}
+
+// readExecCommandFromDesktopFile parses the desktop file to get the Exec entry and
+// checks that the BAMF_DESKTOP_FILE_HINT is present and refers to the desktop file.
+func readExecCommandFromDesktopFile(desktopFile string) (string, error) {
+	var launch string
+
+	file, err := os.Open(desktopFile)
+	if err != nil {
+		return launch, err
+	}
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+
+	in_desktop_section := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if line == "[Desktop Entry]" {
+			in_desktop_section = true
+		} else if strings.HasPrefix(line, "[Desktop Action ") {
+			// maybe later we'll add support here
+			in_desktop_section = false
+		} else if strings.HasPrefix(line, "[") {
+			in_desktop_section = false
+		} else if in_desktop_section && strings.HasPrefix(line, "Exec=") {
+			launch = strings.TrimPrefix(line, "Exec=")
+			break
+		}
+	}
+
+	expectedPrefix := fmt.Sprintf("env BAMF_DESKTOP_FILE_HINT=%s /snap/bin/", desktopFile)
+	if !strings.HasPrefix(launch, expectedPrefix) {
+		return "", fmt.Errorf("Desktop file %q has an unsupported 'Exec' value: %q", desktopFile, launch)
+	}
+
+	return launch, nil
+}
+
+// Parse the Exec command by stripping any exec variables.
+// Passing exec variables (eg, %foo) between confined snaps is unsupported. Currently,
+// we do not have support for passing them in the D-Bus API but there are security
+// implications that must be thought through regarding the influence of the launching
+// snap over the launcher wrt exec variables. For now we simply filter them out.
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+func parseExecCommand(exec_command string) ([]string, error) {
+	args, err := shlex.Split(exec_command)
+	if err != nil {
+		return []string{}, err
+	}
+
+	i := 0
+	for {
+		// We want to keep literal '%' (expressed as '%%') but filter our exec variables
+		// like '%foo'
+		if strings.HasPrefix(args[i], "%%") {
+			args[i] = strings.TrimPrefix(args[i], "%")
+			i++
+		} else if strings.HasPrefix(args[i], "%") {
+			switch args[i] {
+			case "%f", "%F", "%u", "%U", "%i":
+				args = append(args[:i], args[i+1:]...)
+			default:
+				return []string{}, fmt.Errorf("cannot run %q", exec_command)
```

Comment:
Can this error message please include the reason? ie. something like `fmt.Errorf("cannot run %q due to use of %q", exec_command, args[i])`